napari/plugins/_hookexec.py

```diff
@@ -0,0 +1,131 @@
+                except KeyError:
+                    for argname in hook_impl.argnames:
+                        if argname not in caller_kwargs:
+                            raise HookCallError(
+                                "hook call must provide argument %r"
+                                % (argname,)
+                            )
+
+                if hook_impl.hookwrapper:
+                    try:
+                        gen = hook_impl.function(*args)
+                        next(gen)  # first yield
+                        teardowns.append(gen)
+                    except StopIteration:
+                        _raise_wrapfail(gen, "did not yield")
+                else:
+                    res = None
+                    # this is where the plugin function actually gets called
+                    # we put it in a try/except so that if one plugin throws
+                    # an exception, we don't lose the whole loop
+                    try:
+                        res = hook_impl.function(*args)
+                    except Exception as exc:
+                        msg = (
+                            f"Error in plugin '{hook_impl.plugin_name}', "
+                            f"hook '{str(hook_impl.function.__name__)}'"
+                        )
+                        err = PluginError(
+                            msg,
+                            hook_impl.plugin_name,
+                            hook_impl.plugin.__name__,
+                        )
+                        err.__cause__ = exc
+                        # TODO: storing and retrieving these plugin errors is
+                        # being addressed in #1024:
+                        # https://github.com/napari/napari/pull/1024
+
+                    if res is not None:
+                        results.append(res)
+                        if with_impl:
+                            impl_hits.append(hook_impl)
+                        if firstresult:  # halt further impl calls
+                            break
+        except BaseException:
+            excinfo = sys.exc_info()
+    finally:
+        if firstresult:  # first result hooks return a single value
+            outcome = _Result(results[0] if results else None, excinfo)
+        else:
+            outcome = _Result(results, excinfo)
+
+        # run all wrapper post-yield blocks
+        for gen in reversed(teardowns):
+            try:
+                gen.send(outcome)
+                _raise_wrapfail(gen, "has second yield")
+            except StopIteration:
+                pass
+
+        if with_impl:
+            if firstresult:
+                return (
+                    outcome.get_result(),
+                    impl_hits[0] if impl_hits else None,
+                )
+            return list(zip(outcome.get_result(), impl_hits))
+        return outcome.get_result()
+
+
+def _hookexec(
+    hook, skip_imps: Optional[Sequence[HookImpl]] = None, **kwargs
+) -> Any:
```

Comment:
@tlambert03 I feel like this needs a docstring as well... Maybe it can be incorporated into your PR? =)

Reply:
will add.  I feel like that PR might sit there for quite a while, so let's just own it.